### ISSUE TITLE
[BUG]: Timeouts(?) when setting up integration

### ISSUE THREAD

**Issue by dbrgn** [CoMPaTech/stromer#132]
### Describe the bug.

When setting up the integration, I get an error message: "Unknown error occurred"

![image](https://github.com/user-attachments/assets/76046f06-14ba-4eba-b3c8-694f8a63b68c)

The integration is then semi-configured:

![image](https://github.com/user-attachments/assets/5f693a1c-7b30-4a1c-b40f-9c799a0a7120)

In the log, I get this traceback:

```
2024-08-18 18:08:05.618 ERROR (MainThread) [homeassistant.config_entries] Error setting up entry ST3 for stromer
Traceback (most recent call last):
  File "/usr/src/homeassistant/homeassistant/config_entries.py", line 604, in async_setup
    result = await component.async_setup_entry(hass, self)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/config/custom_components/stromer/__init__.py", line 51, in async_setup_entry
    await coordinator.async_config_entry_first_refresh()
  File "/usr/src/homeassistant/homeassistant/helpers/update_coordinator.py", line 281, in async_config_entry_first_refresh
    await self._async_refresh(
  File "/usr/src/homeassistant/homeassistant/helpers/update_coordinator.py", line 354, in _async_refresh
    self.data = await self._async_update_data()
                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/config/custom_components/stromer/coordinator.py", line 31, in _async_update_data
    await self.stromer.stromer_update()
  File "/config/custom_components/stromer/stromer.py", line 84, in stromer_update
    self.status = await self.stromer_call_api(endpoint=endpoint)
                  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/config/custom_components/stromer/stromer.py", line 216, in stromer_call_api
    res = await self._websession.get(url, headers=headers, data={})
          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.12/site-packages/aiohttp/client.py", line 684, in _request
    await resp.start(conn)
  File "/usr/local/lib/python3.12/site-packages/aiohttp/client_reqrep.py", line 999, in start
    message, payload = await protocol.read()  # type: ignore[union-attr]
                       ^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.12/site-packages/aiohttp/streams.py", line 640, in read
    await self._waiter
asyncio.exceptions.CancelledError
```

I suspect some kind of timeout. Maybe the timeout can be increased to avoid the `CancelledError`?

I did not enable debug logs yet. Let me know if that would be helpful.

### What version of Home Assistant Core has the issue?

2024.8.2

### What type of installation are you running?

Home Assistant Supervised

### How did you install the stromer integration

Manually installed `custom_component`

**Comment by dbrgn** [comment]
I dug in a bit more. Here's a (partially redacted) debug log:

```
2024-08-18 19:28:14.483 DEBUG (MainThread) [custom_components.stromer.stromer] Stromer attempt: 1/10
2024-08-18 19:28:14.573 DEBUG (MainThread) [custom_components.stromer.stromer] API call status: 200
2024-08-18 19:28:14.574 DEBUG (MainThread) [custom_components.stromer.stromer] API call returns: {'result': 'ok', 'data': [{'nickname': 'ST3', ...}]}
2024-08-18 19:28:14.574 DEBUG (MainThread) [custom_components.stromer.stromer] Stromer bike: {'nickname': 'ST3', ...}
2024-08-18 19:28:25.800 DEBUG (MainThread) [custom_components.stromer.stromer] API call status: 200
2024-08-18 19:28:25.800 DEBUG (MainThread) [custom_components.stromer.stromer] API call returns: {'result': 'not reachable', 'data': [{'trip_distance': 20.4, ...}]}
2024-08-18 19:28:25.801 DEBUG (MainThread) [custom_components.stromer.stromer] Stromer status: {'trip_distance': 20.4, ...}
2024-08-18 19:28:37.024 DEBUG (MainThread) [custom_components.stromer.stromer] API call status: 200
2024-08-18 19:28:37.024 DEBUG (MainThread) [custom_components.stromer.stromer] API call returns: {'result': 'not reachable', 'data': [{'latitude': 47.0, 'longitude': 8.0, ...}]}
2024-08-18 19:28:37.025 DEBUG (MainThread) [custom_components.stromer.stromer] Stromer position: {'latitude': 47.0, 'longitude': 8.0, ...}
2024-08-18 19:28:37.025 DEBUG (MainThread) [custom_components.stromer.stromer] Stromer connected!
2024-08-18 19:28:37.026 ERROR (MainThread) [homeassistant] Error doing job: Unclosed client session (None)
2024-08-18 19:28:37.029 ERROR (MainThread) [homeassistant] Error doing job: Unclosed connector (None)
2024-08-18 19:28:37.592 DEBUG (MainThread) [custom_components.stromer.stromer] Stromer attempt: 1/10
2024-08-18 19:28:37.658 DEBUG (MainThread) [custom_components.stromer.stromer] API call status: 200
2024-08-18 19:28:37.659 DEBUG (MainThread) [custom_components.stromer.stromer] API call returns: {'result': 'ok', 'data': [{'nickname': 'ST3', ...}]}
2024-08-18 19:28:37.659 DEBUG (MainThread) [custom_components.stromer.stromer] Stromer bike: {'nickname': 'ST3', ...}
2024-08-18 19:28:48.900 DEBUG (MainThread) [custom_components.stromer.stromer] API call status: 200
2024-08-18 19:28:48.901 DEBUG (MainThread) [custom_components.stromer.stromer] API call returns: {'result': 'not reachable', 'data': [{'trip_distance': 20.4,...}]}
2024-08-18 19:28:48.901 DEBUG (MainThread) [custom_components.stromer.stromer] Stromer status: {'trip_distance': 20.4, ...}
2024-08-18 19:29:00.123 DEBUG (MainThread) [custom_components.stromer.stromer] API call status: 200
2024-08-18 19:29:00.123 DEBUG (MainThread) [custom_components.stromer.stromer] API call returns: {'result': 'not reachable', 'data': [{'latitude': 47.0, 'longitude': 8.0, ...}]}
2024-08-18 19:29:00.124 DEBUG (MainThread) [custom_components.stromer.stromer] Stromer position: {'latitude': 47.0, 'longitude': 8.0, ...}
2024-08-18 19:29:00.124 DEBUG (MainThread) [custom_components.stromer.stromer] Stromer connected!
2024-08-18 19:29:00.124 DEBUG (MainThread) [custom_components.stromer.const] Stromer entry: <ConfigEntry entry_id=01J5K8H9D74M4XGGEV0DHQP0XN version=1 domain=stromer title=ST3 state=ConfigEntryState.SETUP_IN_PROGRESS unique_id=stromerbike>
2024-08-18 19:29:00.125 DEBUG (MainThread) [custom_components.stromer.stromer] Stromer attempt: 1/10
2024-08-18 19:29:00.212 DEBUG (MainThread) [custom_components.stromer.stromer] API call status: 200
2024-08-18 19:29:00.213 DEBUG (MainThread) [custom_components.stromer.stromer] API call returns: {'result': 'ok', 'data': [{'nickname': 'ST3', ...}]}
2024-08-18 19:29:00.213 DEBUG (MainThread) [custom_components.stromer.stromer] Stromer bike: {'nickname': 'ST3', ...}
2024-08-18 19:29:11.422 DEBUG (MainThread) [custom_components.stromer.stromer] API call status: 200
2024-08-18 19:29:11.423 DEBUG (MainThread) [custom_components.stromer.stromer] API call returns: {'result': 'not reachable', 'data': [{'trip_distance': 20.4, ...}]}
2024-08-18 19:29:11.424 DEBUG (MainThread) [custom_components.stromer.stromer] Stromer status: {'trip_distance': 20.4, ...}
2024-08-18 19:29:13.914 DEBUG (MainThread) [custom_components.stromer.const] Finished fetching stromer data in 13.789 seconds (success: True)
2024-08-18 19:29:13.915 ERROR (MainThread) [homeassistant.config_entries] Error setting up entry ST3 for stromer
Traceback (most recent call last):
File "/usr/src/homeassistant/homeassistant/config_entries.py", line 604, in async_setup
result = await component.async_setup_entry(hass, self)
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
File "/config/custom_components/stromer/__init__.py", line 51, in async_setup_entry
await coordinator.async_config_entry_first_refresh()
File "/usr/src/homeassistant/homeassistant/helpers/update_coordinator.py", line 281, in async_config_entry_first_refresh
await self._async_refresh(
File "/usr/src/homeassistant/homeassistant/helpers/update_coordinator.py", line 354, in _async_refresh
self.data = await self._async_update_data()
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
File "/config/custom_components/stromer/coordinator.py", line 31, in _async_update_data
await self.stromer.stromer_update()
File "/config/custom_components/stromer/stromer.py", line 89, in stromer_update
self.position = await self.stromer_call_api(endpoint=endpoint)
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
File "/config/custom_components/stromer/stromer.py", line 216, in stromer_call_api
res = await self._websession.get(url, headers=headers, data={})
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
File "/usr/local/lib/python3.12/site-packages/aiohttp/client.py", line 684, in _request
await resp.start(conn)
File "/usr/local/lib/python3.12/site-packages/aiohttp/client_reqrep.py", line 999, in start
message, payload = await protocol.read() # type: ignore[union-attr]
^^^^^^^^^^^^^^^^^^^^^
File "/usr/local/lib/python3.12/site-packages/aiohttp/streams.py", line 640, in read
await self._waiter
asyncio.exceptions.CancelledError
```

Data can be fetched, but the integration still fails. Maybe related to these log entries?

> 2024-08-18 19:28:37.026 ERROR (MainThread) [homeassistant] Error doing job: Unclosed client session (None)
> 2024-08-18 19:28:37.029 ERROR (MainThread) [homeassistant] Error doing job: Unclosed connector (None)

**Comment by CoMPaTech** [comment]
Hi @dbrgn - I would need to check why this happens, it sort of looks like the API (of Stromer) isn't returning nicely. Good job to redact properly! No headspace available tonight unfortunately, I'll look into it in the morning. Assuming by install manually you just used `main` to deploy? Or did you specifically use a release/tag?

Looks like you get past the initial connection (i.e. to detect your stromer (actually, if using main/v4 - to determine if there is more than one stromer). But once that gets handed over to the coordinator (which is responsible for keeping it up to date) it doesn't reply in time).

I've not seen the HA notices myself, but the code definitely needs some TLC on properly closing connections. I'll try to reproduce on my side as well, please let me know how you deployed locally (as in git clone using main or which version). Tnx!

**Comment by dbrgn** [comment]
Hi @CoMPaTech, thanks for the quick reply! I cloned `main` directly and symlinked `stromer/custom_components/stromer` into my custom components directory.

Regarding the unclosed connections, I found the reason while digging in the code. The aiohttp session is opened twice (once for validating the credentials, once for actually fetching data) and never closed. I'll submit a PR that fixes the issue.

I also saw that the default timeout in the `Stromer` class is 60s and increased it to 180s. Unfortunately with both changes the issue still persists. Here's another log.

```
2024-08-18 19:55:59.063 DEBUG (MainThread) [custom_components.stromer.stromer] Initializing Stromer with API version v4 and timeout 180
2024-08-18 19:55:59.746 DEBUG (MainThread) [custom_components.stromer.stromer] Stromer attempt: 1/10
2024-08-18 19:55:59.825 DEBUG (MainThread) [custom_components.stromer.stromer] API call status: 200
...
2024-08-18 19:56:22.318 DEBUG (MainThread) [custom_components.stromer.stromer] Stromer connected!
2024-08-18 19:56:22.318 DEBUG (MainThread) [custom_components.stromer.config_flow] Credentials validated successfully
2024-08-18 19:56:22.318 DEBUG (MainThread) [custom_components.stromer.stromer] Closing aiohttp session
2024-08-18 19:56:22.328 DEBUG (MainThread) [custom_components.stromer.stromer] Initializing Stromer with API version v4 and timeout 180
2024-08-18 19:56:22.917 DEBUG (MainThread) [custom_components.stromer.stromer] Stromer attempt: 1/10
2024-08-18 19:56:23.048 DEBUG (MainThread) [custom_components.stromer.stromer] API call status: 200
...
2024-08-18 19:56:45.534 DEBUG (MainThread) [custom_components.stromer.stromer] Stromer connected!
2024-08-18 19:56:45.535 DEBUG (MainThread) [custom_components.stromer.const] Stromer entry: <ConfigEntry entry_id=01J5KA43NP10M8KT7W2CTGBMY1 version=1 domain=stromer title=ST3 state=ConfigEntryState.SETUP_IN_PROGRESS unique_id=stromerbike>
2024-08-18 19:56:45.535 DEBUG (MainThread) [custom_components.stromer.stromer] Stromer attempt: 1/10
2024-08-18 19:56:45.645 DEBUG (MainThread) [custom_components.stromer.stromer] API call status: 200
...
2024-08-18 19:56:59.061 DEBUG (MainThread) [custom_components.stromer.const] Finished fetching stromer data in 13.526 seconds (success: True)
2024-08-18 19:56:59.062 ERROR (MainThread) [homeassistant.config_entries] Error setting up entry ST3 for stromer
Traceback (most recent call last):
File "/usr/src/homeassistant/homeassistant/config_entries.py", line 604, in async_setup
result = await component.async_setup_entry(hass, self)
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
File "/config/custom_components/stromer/__init__.py", line 51, in async_setup_entry
await coordinator.async_config_entry_first_refresh()
File "/usr/src/homeassistant/homeassistant/helpers/update_coordinator.py", line 281, in async_config_entry_first_refresh
await self._async_refresh(
File "/usr/src/homeassistant/homeassistant/helpers/update_coordinator.py", line 354, in _async_refresh
self.data = await self._async_update_data()
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
File "/config/custom_components/stromer/coordinator.py", line 31, in _async_update_data
await self.stromer.stromer_update()
File "/config/custom_components/stromer/stromer.py", line 96, in stromer_update
self.position = await self.stromer_call_api(endpoint=endpoint)
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
File "/config/custom_components/stromer/stromer.py", line 223, in stromer_call_api
res = await self._websession.get(url, headers=headers, data={})
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
File "/usr/local/lib/python3.12/site-packages/aiohttp/client.py", line 684, in _request
await resp.start(conn)
File "/usr/local/lib/python3.12/site-packages/aiohttp/client_reqrep.py", line 999, in start
message, payload = await protocol.read() # type: ignore[union-attr]
^^^^^^^^^^^^^^^^^^^^^
File "/usr/local/lib/python3.12/site-packages/aiohttp/streams.py", line 640, in read
await self._waiter
asyncio.exceptions.CancelledError
```

The suspect thing is that in every case in the logs I watched, the exception popped up exactly after 60 seconds. In this case it started at 19:55:59.063 and failed at 19:56:59.062. However, it can't be the aiohttp connection, because that is opened, then closed, and then re-opened, and definitely not yet open for 60 seconds.

Does the config flow itself have a timeout? I'll dig in some more.

**Comment by CoMPaTech** [comment]
Main still doesn't have the `morebikes` complexity in there (while also some improvements, but nothing related to proper session closure). Just for trying out, could you give `morebikes` a quick try to see what that one does (e.g. if it also get's to the coordinator or fail earlier)? The handling in that `stromer.py` somewhat changed, I'm kind of curious if it fails earlier wrt another call in between.

**Comment by dbrgn** [comment]
@CoMPaTech the commit with the connection closing is here (created against `main`): https://github.com/CoMPaTech/stromer/compare/main...dbrgn:homeassistant-stromer:close-connections I also had ideas to improve it, by making the `Stromer` class a context manager. However, it seems that `morebikes` is the active branch and I probably shouldn't make PRs against `main` right now? Do you accept PRs against your feature branch?

I haven't tried the `morebikes` branch _yet_, but I have another datapoint: My production homeassistant node is running on a SBC with not a lot of computation power. When I ran the integration on my regular development computer, the integration setup worked fine at first attempt. Initially I thought that it was simply faster than 60 seconds, but in reality it took 80s. I'm a bit confused now why it works on my dev setup, but not on my production setup.

**Comment by dbrgn** [comment]
@CoMPaTech I tried the `morebikes` branch now, and it worked flawlessly from the start, very cool!

I think by splitting up the process, there is less chance of running into some kind of timeout. Once #82 is ready, this issue can probably be considered resolved.

I will create a pull request with the small improvements against your branch, then you can decide what to do with it.

Regarding the state of #82, how "ready" do you think it is? Are there known issues with the branch?

**Comment by CoMPaTech** [comment]
To combine two in one - I started `morebikes` mostly because I only have one and would need others to test the waters. The recent changes to it though were to also have 'existing' users still being able to use if properly. As in: don't ditch the data before/after new version. So it's actually trifold; existing single user users migration, new (single) user adoption and existing multi-bike users ... with two out of three ticket off since you tested the second condition I would be confident to just 'merge' and release to beta (instead of alpha) so normal development can continue.

Committing against the branch is OK with me. Thanks for the PR, let me know if we can already merge it once the actions have run etc.

W.r.t. timing that indeed is weird since it should all be relatively small calls and I presume it's mostly waiting on the API. I'm also puzzled by why the `morebikes` would work (effectively it's an additional call in between).

Best approach might be to merge your #134 - test it on my side as well, than merge #82 and (where required) release to beta so we at least have a clean main to work from. Feedback from the lucky users with multiple Stromers can follow from there. (For reference, those are included/mentioned in #81)

**Comment by CoMPaTech** [comment]
I merged #134 and one tiny commit to make CodeRabbitAI (more) happy, released as v0.4.0b1 (against more bikes). As included in #81 and the release, if no problematic feedback I would just release #82 to main with a quick run as b2 and then to release. (Especially since HACS doesn't really play nicely from 2.0 onward showing the option to (not) beta download)

**Comment by dbrgn** [comment]
> W.r.t. timing that indeed is weird since it should all be relatively small calls and I presume it's mostly waiting on the API. I'm also puzzled by why the `morebikes` would work (effectively it's an additional call in between).

The difference is that it is now a two-stage process: First the bikes are queried, then the user has to select a bike. And only then the actual registration of the device happens. Because there are now two shorter processes with an intermediate state, and not one single long asynchronous process, that might be the reason why it works now.

Thanks for the update, I'll upgrade to the beta tag soon and will report any issues!

**Comment by dbrgn** [comment]
I think this can be closed for now, since #81 is merged? Thanks!

**Comment by CoMPaTech** [comment]
All tnx and kudos to you for testing and contributing :)
